pinot-core/src/main/java/com/linkedin/pinot/core/startree/StarTreeV2.java

```diff
@@ -0,0 +1,226 @@
+    if (rootNodeOffset > size) {
+      dataBuffer.close();
+      dataBuffer = new MMapBuffer(starTreeFile, 0, rootNodeOffset, MMapMode.READ_ONLY);
+    }
+
+    int numDimensions = dataBuffer.getInt(offset);
+    offset += Integer.BYTES;
+
+    dimensionNameToIndexMap = HashBiMap.create(numDimensions);
+    byte[] dimensionNameBytes = new byte[DIMENSION_NAME_MAX_LENGTH];
+
+    for (int i = 0; i < numDimensions; i++) {
+      int index = dataBuffer.getInt(offset);
+      offset += Integer.BYTES;
+
+      int dimensionLength = dataBuffer.getInt(offset);
+      offset += Integer.BYTES;
+
+      // Since we are re-using the same bytes for reading strings, assert we have allocated enough.
+      Preconditions.checkState(dimensionLength < DIMENSION_NAME_MAX_LENGTH);
+
+      // Ok to cast offset to int, as its value is too small at this point in the file.
+      dataBuffer.copyTo((int) offset, dimensionNameBytes, 0, dimensionLength);
+      offset += dimensionLength;
+
+      String dimensionName = new String(dimensionNameBytes, 0, dimensionLength, UTF8);
+      dimensionNameToIndexMap.put(dimensionName, index);
+    }
+
+    numNodes = dataBuffer.getInt(offset);
+    offset += Integer.BYTES;
+
+    Preconditions.checkState((offset == rootNodeOffset), "Error reading Star Tree file, header length mis-match");
+
+    dataBuffer.close();
+    return offset;
+  }
+
+  /**
+   * Helper method that loads the star tree into a LBuffer, from
+   * the given file.
+   *
+   * @param starTreeFile
+   * @param rootOffset
+   * @return
+   * @throws IOException
+   */
+  private static LBufferAPI loadFromFile(File starTreeFile, long rootOffset)
+      throws IOException {
+    FileChannel fileChannel = new FileInputStream(starTreeFile).getChannel();
+    ByteBuffer byteBuffer = ByteBuffer.allocate(BUFFER_SIZE);
+
+    long start = rootOffset;
+    long end = starTreeFile.length();
+    long destOffset = 0;
+    LBufferAPI lBuffer = new LBuffer(end - start);
+
+    int bytesRead = 0;
+    while ((bytesRead = fileChannel.read(byteBuffer, start)) > 0) {
+      lBuffer.readFrom(byteBuffer.array(), 0, destOffset, bytesRead);
+      start += bytesRead;
+      destOffset += bytesRead;
+      byteBuffer.clear();
+    }
+    fileChannel.close();
+    return lBuffer;
+  }
+
+  /**
+   * {@inheritDoc}
+   * @return
+   */
+  @Override
+  public StarTreeIndexNodeInterf getRoot() {
+    return root;
+  }
+
+  /**
+   * {@inheritDoc}
+   * @return
+   */
+  public Version getVersion() {
+    return Version.V2;
+  }
+
+  @Override
+  public int getNumNodes() {
+    return numNodes;
+  }
+
+  /**
+   * {@inheritDoc}
+   * @return
+   */
+  @Override
+  public HashBiMap<String, Integer> getDimensionNameToIndexMap() {
+    return dimensionNameToIndexMap;
+  }
+
+  /**
+   * {@inheritDoc}
+   * @param outputFile
+   * @throws IOException
+   */
+  @Override
+  public void writeTree(File outputFile)
+      throws IOException {
+
```

Comment:
I added an exception for now, since I don't know if (how) this would be useful. Will revisit it with the next set of changes.